core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala

```diff
@@ -290,15 +310,17 @@ private[spark] class ExecutorMonitor(
   }
 
   override def onExecutorAdded(event: SparkListenerExecutorAdded): Unit = {
-    val exec = ensureExecutorIsTracked(event.executorId)
+    val exec = ensureExecutorIsTracked(event.executorId, event.executorInfo.resourceProfileId)
     exec.updateRunningTasks(0)
     logInfo(s"New executor ${event.executorId} has registered (new total is ${executors.size()})")
   }
 
   override def onExecutorRemoved(event: SparkListenerExecutorRemoved): Unit = {
     val removed = executors.remove(event.executorId)
     if (removed != null) {
-      logInfo(s"Executor ${event.executorId} removed (new total is ${executors.size()})")
+      val count = execResourceProfileCount.getOrDefault(removed.resourceProfileId, 0)
+      execResourceProfileCount.replace(removed.resourceProfileId, count, count - 1)
+      execResourceProfileCount.remove(removed.resourceProfileId, 0)
```

Comment:
Minor but this could be a small helper method (since it's basically the same code as later when you decrement the number of unknowns).